File: scripts/backend-python/payment_link_service.py
"""Create Payment Link - POST /link/create.

Doc: https://www.paytmpayments.com/docs/api/create-link-api
Defaults & gotchas baked in:
- head requires tokenType: "AES" + timestamp (Unix epoch SECONDS as string)
- linkType: "FIXED" by default (GENERIC ignores amount)
- amount is a JSON number, NOT a string
- linkName AND linkDescription must each be >= 3 chars, alphanumerics + spaces
  only (same rule for both fields). Sanitize via _sanitize_description() below.
- customer details nested under customerContact (not top-level)
- expiryDate format DD/MM/YYYY HH:MM:SS (most MIDs)
"""
import json
import re
import secrets
from datetime import datetime, timedelta, timezone
from decimal import Decimal, ROUND_HALF_UP
from typing import Any, Optional

import requests
from paytmchecksum import PaytmChecksum

from paytm_config import get_paytm_config
from paytm_service import PaytmError

IST = timezone(timedelta(hours=5, minutes=30))


def _sanitize_description(s: Optional[str], fallback: str = "Invoice payment") -> str:
    cleaned = re.sub(r"\s+", " ", re.sub(r"[^A-Za-z0-9 ]", " ", s or "")).strip()
    return cleaned if len(cleaned) >= 3 else fallback


def _expiry_one_year_from_now() -> str:
    d = datetime.now(IST) + timedelta(days=365)
    return d.strftime("%d/%m/%Y 23:59:59")


_TWO_PLACES = Decimal("0.01")
_FALLBACK = Decimal("1.00")


def _normalize_amount(amount: Any) -> float:
    """Two-decimal currency normalization using Decimal to avoid binary-float drift.

    Paytm's Create Link API requires `amount` as a JSON number (not a string),
    so we must emit a float at the end - but we do all the rounding in Decimal
    space first, then convert at the very last step. For typical INR amounts
    this gives bit-exact two-decimal precision; the float conversion only loses
    significance well above the rupee-amount range any real merchant will use.
    """
    raw = "" if amount is None else str(amount).strip()
    try:
        d = Decimal(raw) if raw else _FALLBACK
    except Exception:
        d = _FALLBACK
    if d <= 0:
        d = _FALLBACK
    d = d.quantize(_TWO_PLACES, rounding=ROUND_HALF_UP)
    return float(d)


def create_payment_link(
    *,
    amount: Any = "1.00",
    link_name: Optional[str] = None,
    link_description: Optional[str] = None,
    customer_name: Optional[str] = None,
    customer_email: Optional[str] = None,
    customer_mobile: Optional[str] = None,
    customer_id: Optional[str] = None,
    expiry_date: Optional[str] = None,
    order_id: Optional[str] = None,
    send_sms: bool = True,
    send_email: bool = True,
    callback_url: Optional[str] = None,
    merchant_unique_reference: Optional[str] = None,
    server_base_url: str = "",
) -> dict:
    cfg = get_paytm_config()
    if not cfg["mid"]:
        raise PaytmError("MISSING_MID", "Missing PAYTM_MID")
    if not cfg["merchant_key"]:
        raise PaytmError("MISSING_MERCHANT_KEY", "Missing PAYTM_MERCHANT_KEY")

    order_id = (order_id or "").strip() or ("LNK_" + secrets.token_hex(10).upper())
    callback_url = (callback_url or cfg["callback_url"]
                    or f"{server_base_url.rstrip('/')}/paytm/callback")

    customer_contact: dict = {}
    if customer_name and customer_name.strip(): customer_contact["customerName"] = customer_name.strip()
    if customer_email and customer_email.strip(): customer_contact["customerEmail"] = customer_email.strip()
    if customer_mobile and customer_mobile.strip(): customer_contact["customerMobile"] = customer_mobile.strip()
    if customer_id and customer_id.strip(): customer_contact["customerId"] = customer_id.strip()

    body: dict = {
        "mid": cfg["mid"],
        "linkType": "FIXED",
        "linkName": _sanitize_description(link_name, "Invoice"),
        "linkDescription": _sanitize_description(link_description, "Invoice payment"),
        "amount": _normalize_amount(amount),
        "sendSms": bool(send_sms),
        "sendEmail": bool(send_email),
        "customerContact": customer_contact,
        "expiryDate": (expiry_date or "").strip() or _expiry_one_year_from_now(),
        "orderId": order_id,
        "callbackUrl": callback_url,
    }
    if merchant_unique_reference and merchant_unique_reference.strip():
        body["merchantUniqueReference"] = merchant_unique_reference.strip()

    signature = PaytmChecksum.generateSignature(json.dumps(body), cfg["merchant_key"])
    head = {
        "tokenType": "AES",
        "signature": signature,
        "timestamp": str(int(datetime.now(timezone.utc).timestamp())),
    }

    r = requests.post(cfg["link_create_url"], json={"head": head, "body": body},
                      headers={"Content-Type": "application/json"}, timeout=15)
    text = r.text
    if not r.ok:
        raise _upstream("LINK_HTTP_ERROR", f"link/create HTTP {r.status_code}", order_id, text)

    data = json.loads(text)
    info = (data.get("body") or {}).get("resultInfo") or {}
    status = info.get("resultStatus")
    if status and status not in ("SUCCESS", "S"):
        raise _upstream("LINK_FAILED", info.get("resultMsg") or "link/create failed", order_id, text)

    body_resp = data.get("body") or {}
    # Read defensively - current Paytm returns linkId; older docs LinkID.
    link_id = body_resp.get("linkId") if body_resp.get("linkId") is not None else body_resp.get("LinkID")

    return {
        "orderId": order_id,
        "linkId": link_id,
        "shortUrl": body_resp.get("shortUrl"),
        "longUrl": body_resp.get("longUrl"),
        "linkStatus": body_resp.get("linkStatus") or "ACTIVE",
        "amount": body["amount"],
        "mid": cfg["mid"],
    }


def fetch_link_transactions(
    *,
    link_id: Any,
    page_no: int = 1,
    page_size: int = 10,
    fetch_all_txns: bool = True,
) -> dict:
    """Fetch transactions for a Payment Link - POST /link/fetchTransaction.

    Doc: https://www.paytmpayments.com/docs/api/fetch-transaction-link-api

    USE THIS for Payment Link reconciliation, NOT /v3/order/status. Returns
    {linkId, orders: [...], resultInfo}. A 404 from Paytm ("Data Not Found")
    is normalised to an empty list so callers don't need to special-case
    "not paid yet".
    """
    cfg = get_paytm_config()
    if not cfg["mid"]:
        raise PaytmError("MISSING_MID", "Missing PAYTM_MID")
    if not cfg["merchant_key"]:
        raise PaytmError("MISSING_MERCHANT_KEY", "Missing PAYTM_MERCHANT_KEY")
    if link_id in (None, ""):
        raise PaytmError("MISSING_LINK_ID", "linkId is required")

    # linkId MUST be a JSON number - coerce defensively.
    try:
        numeric_link_id = int(link_id)
    except (TypeError, ValueError):
        raise PaytmError("INVALID_LINK_ID", f"linkId must be numeric, got: {link_id!r}")

    body = {
        "mid": cfg["mid"],
        "linkId": numeric_link_id,
        "pageNo": int(page_no) or 1,
        "pageSize": int(page_size) or 10,
        "fetchAllTxns": bool(fetch_all_txns),
    }
    signature = PaytmChecksum.generateSignature(json.dumps(body), cfg["merchant_key"])
    head = {"tokenType": "AES", "signature": signature}

    r = requests.post(
        cfg["link_fetch_transaction_url"],
        json={"head": head, "body": body},
        headers={"Content-Type": "application/json"},
        timeout=15,
    )
    text = r.text
    if not r.ok:
        raise _upstream(
            "LINK_FETCH_TXN_HTTP_ERROR",
            f"link/fetchTransaction HTTP {r.status_code}",
            str(numeric_link_id),
            text,
        )

    data = json.loads(text)
    info = (data.get("body") or {}).get("resultInfo") or {}
    msg = info.get("resultMessage") or info.get("resultMsg") or ""
    # 404 = "Data Not Found" - link exists, no txns yet. Normalise to empty list.
    if info.get("resultCode") == "404" or "not found" in msg.lower():
        return {"linkId": numeric_link_id, "orders": [], "resultInfo": info}
    status = info.get("resultStatus")
    if status and status not in ("SUCCESS", "S"):
        raise _upstream(
            "LINK_FETCH_TXN_FAILED",
            msg or "link/fetchTransaction failed",
            str(numeric_link_id),
            text,
        )
    return {
        "linkId": numeric_link_id,
        "orders": (data.get("body") or {}).get("orders") or [],
        "resultInfo": info,
    }


def _upstream(code: str, message: str, order_id: str, raw: str) -> PaytmError:
    paytm: dict = {}
    try:
        info = ((json.loads(raw) or {}).get("body") or {}).get("resultInfo") or {}
        for k in ("resultStatus", "resultCode", "resultMsg"):
            if info.get(k):
                paytm[k] = info[k]
    except Exception:
        pass
    return PaytmError(code, message, http_status=502, order_id=order_id, paytm=paytm or None)

File: scripts/backend-node/paymentLinkService.js
// Create Payment Link - POST /link/create
// Doc: https://www.paytmpayments.com/docs/api/create-link-api
// Defaults & gotchas baked in:
//   - head requires tokenType: "AES" + timestamp (Unix epoch SECONDS as string)
//   - linkType: "FIXED" by default (GENERIC ignores amount)
//   - amount is a JSON number, NOT a string
//   - linkName: alphanumerics ONLY (no spaces) - some MIDs reject space here.
//     Sanitize via sanitizeLinkName() below.
//   - linkDescription: alphanumerics + spaces. Sanitize via sanitizeDescription().
//     Both fields must be ≥ 3 chars.
//   - customer details nested under customerContact (not top-level)
//   - expiryDate format DD/MM/YYYY HH:MM:SS (most MIDs)
import crypto from "node:crypto";
import PaytmChecksum from "paytmchecksum";
import { getPaytmConfig } from "./paytmConfig.js";

// linkDescription: alphanumerics + spaces (per Paytm docs). Sanitize untrusted
// input; fall back when empty / too short.
function sanitizeDescription(s, fallback = "Invoice payment") {
  const cleaned = String(s ?? "").replace(/[^A-Za-z0-9 ]/g, " ").replace(/\s+/g, " ").trim();
  return cleaned.length >= 3 ? cleaned : fallback;
}

// linkName: alphanumerics ONLY. Paytm's docs say spaces are allowed but several
// MIDs reject space as a special character with "link name contains special
// character". Stripping spaces is the safe cross-MID default.
function sanitizeLinkName(s, fallback = "Invoice") {
  const cleaned = String(s ?? "").replace(/[^A-Za-z0-9]/g, "");
  return cleaned.length >= 3 ? cleaned : fallback;
}

function expiryOneYearFromNow() {
  // DD/MM/YYYY HH:MM:SS in IST
  const offsetMs = 5.5 * 60 * 60 * 1000;
  const d = new Date(Date.now() + offsetMs + 365 * 24 * 60 * 60 * 1000);
  const pad = n => String(n).padStart(2, "0");
  return `${pad(d.getUTCDate())}/${pad(d.getUTCMonth() + 1)}/${d.getUTCFullYear()} 23:59:59`;
}

function normalizeAmount(amount) {
  // Payment Link API takes `amount` as a JSON number (not a string).
  //
  // To stay precision-safe we do the rounding via integer paise (×100, round to
  // nearest, ÷100) instead of `Number.toFixed()` directly - this guarantees
  // bit-exact two-decimal currency math without depending on string conversion
  // for the rounding step. Float conversion only happens at the final ÷100, so
  // typical INR amounts are exact (the float error budget kicks in well above
  // any rupee amount a real merchant will send).
  //
  // For larger / more sensitive currency math, swap this for a decimal library
  // (decimal.js, big.js) and emit via JSON.stringify with a number coercion.
  const raw = String(amount ?? "").trim();
  const n = Number(raw);
  if (!Number.isFinite(n) || n <= 0) return 1.00;
  // Math.round gives banker-style rounding via float; for half-up, add Number.EPSILON.
  return Math.round((n + Number.EPSILON) * 100) / 100;
}

export async function createPaymentLink({
  amount,
  linkName,
  linkDescription,
  customerName,
  customerEmail,
  customerMobile,
  customerId,
  expiryDate,
  orderId: callerOrderId,
  sendSms = true,
  sendEmail = true,
  callbackUrl: callerCallback,
  merchantUniqueReference,
  serverBaseUrl,
}) {
  const cfg = getPaytmConfig();
  if (!cfg.mid) throw asError("MISSING_MID", "Missing PAYTM_MID");
  if (!cfg.merchantKey) throw asError("MISSING_MERCHANT_KEY", "Missing PAYTM_MERCHANT_KEY");

  const orderId = callerOrderId?.trim()
    || `LNK_${crypto.randomBytes(10).toString("hex").toUpperCase()}`;
  const callbackUrl = callerCallback?.trim() || cfg.callbackUrl?.trim()
    || `${(serverBaseUrl || cfg.callbackBase).replace(/\/+$/, "")}/paytm/callback`;

  const body = {
    mid: cfg.mid,
    linkType: "FIXED",
    linkName: sanitizeLinkName(linkName, "Invoice"),
    linkDescription: sanitizeDescription(linkDescription, "Invoice payment"),
    amount: normalizeAmount(amount),
    sendSms: !!sendSms,
    sendEmail: !!sendEmail,
    customerContact: {
      ...(customerName?.trim() ? { customerName: customerName.trim() } : {}),
      ...(customerEmail?.trim() ? { customerEmail: customerEmail.trim() } : {}),
      ...(customerMobile?.trim() ? { customerMobile: customerMobile.trim() } : {}),
      ...(customerId?.trim() ? { customerId: customerId.trim() } : {}),
    },
    expiryDate: expiryDate?.trim() || expiryOneYearFromNow(),
    orderId,
    callbackUrl,
    ...(merchantUniqueReference?.trim() ? { merchantUniqueReference: merchantUniqueReference.trim() } : {}),
  };

  const signature = await PaytmChecksum.generateSignature(JSON.stringify(body), cfg.merchantKey);
  const head = {
    tokenType: "AES",
    signature,
    timestamp: String(Math.floor(Date.now() / 1000)),
  };

  const r = await fetch(cfg.linkCreateUrl, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ head, body }),
  });
  const text = await r.text();
  if (!r.ok) throw upstream("LINK_HTTP_ERROR", `link/create HTTP ${r.status}`, orderId, text);

  const json = JSON.parse(text);
  const status = json?.body?.resultInfo?.resultStatus;
  if (status && status !== "SUCCESS" && status !== "S") {
    throw upstream("LINK_FAILED", json?.body?.resultInfo?.resultMsg || "link/create failed", orderId, text);
  }
  // Read defensively - current Paytm returns linkId; older docs LinkID.
  const linkId = json?.body?.linkId ?? json?.body?.LinkID;

  return {
    orderId,
    linkId,
    shortUrl: json?.body?.shortUrl,
    longUrl: json?.body?.longUrl,
    linkStatus: json?.body?.linkStatus || "ACTIVE",
    amount: body.amount,
    mid: cfg.mid,
  };
}

/**
 * Fetch transactions for a Payment Link - POST /link/fetchTransaction.
 * Doc: https://www.paytmpayments.com/docs/api/fetch-transaction-link-api
 *
 * USE THIS for Payment Link reconciliation, NOT /v3/order/status.
 * Returns {orders: [...]} - always an array, may be empty (404 from Paytm
 * is normalised to an empty array here so callers don't have to special-case
 * "not paid yet").
 */
export async function fetchLinkTransactions({
  linkId,
  pageNo = 1,
  pageSize = 10,
  fetchAllTxns = true,
}) {
  const cfg = getPaytmConfig();
  if (!cfg.mid) throw asError("MISSING_MID", "Missing PAYTM_MID");
  if (!cfg.merchantKey) throw asError("MISSING_MERCHANT_KEY", "Missing PAYTM_MERCHANT_KEY");
  if (linkId === undefined || linkId === null || linkId === "") {
    throw asError("MISSING_LINK_ID", "linkId is required");
  }

  // linkId MUST be a JSON number, not a string. Coerce defensively here so
  // callers can pass either form without surprise.
  const numericLinkId = Number(linkId);
  if (!Number.isFinite(numericLinkId)) {
    throw asError("INVALID_LINK_ID", `linkId must be numeric, got: ${linkId}`);
  }

  const body = {
    mid: cfg.mid,
    linkId: numericLinkId,
    pageNo: Number(pageNo) || 1,
    pageSize: Number(pageSize) || 10,
    fetchAllTxns: !!fetchAllTxns,
  };
  const signature = await PaytmChecksum.generateSignature(JSON.stringify(body), cfg.merchantKey);
  const head = { tokenType: "AES", signature };

  const r = await fetch(cfg.linkFetchTransactionUrl, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ head, body }),
  });
  const text = await r.text();
  if (!r.ok) throw upstream("LINK_FETCH_TXN_HTTP_ERROR", `link/fetchTransaction HTTP ${r.status}`, String(numericLinkId), text);

  const json = JSON.parse(text);
  const info = json?.body?.resultInfo;
  // 404 = "Data Not Found" - link exists but no transactions yet. Normalise
  // to an empty list so callers can treat all valid responses uniformly.
  if (info?.resultCode === "404" || /not\s*found/i.test(info?.resultMessage || info?.resultMsg || "")) {
    return { linkId: numericLinkId, orders: [], resultInfo: info };
  }
  if (info?.resultStatus && info.resultStatus !== "SUCCESS" && info.resultStatus !== "S") {
    throw upstream("LINK_FETCH_TXN_FAILED", info.resultMessage || info.resultMsg || "link/fetchTransaction failed",
      String(numericLinkId), text);
  }
  return {
    linkId: numericLinkId,
    orders: json?.body?.orders ?? [],
    resultInfo: info,
  };
}

function asError(code, message) {
  const e = new Error(message);
  e.code = code;
  return e;
}

function upstream(code, message, orderId, raw) {
  const e = new Error(message);
  e.code = code;
  e.httpStatus = 502;
  e.orderId = orderId;
  try {
    const info = JSON.parse(raw)?.body?.resultInfo;
    if (info) e.paytm = { resultStatus: info.resultStatus, resultCode: info.resultCode, resultMsg: info.resultMsg };
  } catch {}
  return e;
}
